trunk/src/app/srs_app_rtsp.cpp

```diff
@@ -0,0 +1,589 @@
+            std::string sdp;
+            if ((err = session_->do_describe(req.get(), sdp)) != srs_success) {
+                res->status = SRS_CONSTS_RTSP_InternalServerError;
+                if (srs_error_code(err) == ERROR_SYSTEM_SECURITY_DENY) {
+                    res->status = SRS_CONSTS_RTSP_Forbidden;
+                }
+                srs_warn("describe failed: %s", srs_error_desc(err).c_str());
+                srs_error_reset(err);
+            }
+
+            res->sdp = sdp;
+            if ((err = rtsp_->send_message(res.get())) != srs_success) {
+                return  srs_error_wrap(err, "response describe");
+            }
+        } else if (req->is_setup()) {
+            srs_assert(req->transport);            
+
+            SrsUniquePtr<SrsRtspSetupResponse> res(new SrsRtspSetupResponse((int)req->seq));
+            res->session = session_id_;
+
+            uint32_t ssrc = 0;
+            if ((err = session_->do_setup(req.get(), &ssrc)) != srs_success) {
+                res->status = SRS_CONSTS_RTSP_InternalServerError;
+                srs_warn("setup failed: %s", srs_error_desc(err).c_str());
+                srs_error_reset(err);
+            }
+    
+            res->transport->copy(req->transport);
+            res->session = session_id_;
+            res->ssrc = srs_int2str(ssrc);
+            res->client_port_min = req->transport->client_port_min;
+            res->client_port_max = req->transport->client_port_max;
+            // TODO: FIXME: listen local port
+            res->local_port_min = 0;
+            res->local_port_max = 0;
+            if ((err = rtsp_->send_message(res.get())) != srs_success) {  
+                return srs_error_wrap(err, "response setup");
+            }
+        } else if (req->is_play()) {
+            SrsUniquePtr<SrsRtspResponse> res(new SrsRtspResponse((int)req->seq));
+            res->session = session_id_;
+            if ((err = rtsp_->send_message(res.get())) != srs_success) {
+                return srs_error_wrap(err, "response record");
+            }
+            
+            if ((err = session_->do_play(req.get(), this)) != srs_success) {
+                return srs_error_wrap(err, "prepare play");
+            }
+        } else if (req->is_teardown()) {
+            SrsUniquePtr<SrsRtspResponse> res(new SrsRtspResponse((int)req->seq));
+            res->session = session_id_;
+            if ((err = rtsp_->send_message(res.get())) != srs_success) {
+                return srs_error_wrap(err, "response teardown");
+            }
+
+            if ((err = session_->do_teardown()) != srs_success) {
+                return srs_error_wrap(err, "teardown");
+            }
+        }
+    }
+    
+    return err;
+}
+
+SrsRtspUdpNetwork::SrsRtspUdpNetwork()
+{
+    addr_ = NULL;
+    stfd_ = NULL;
+}
+
+SrsRtspUdpNetwork::~SrsRtspUdpNetwork()
+{
+    srs_close_stfd(stfd_);
+    srs_freep(addr_);
+}
+
+srs_error_t SrsRtspUdpNetwork::initialize(std::string ip, int port)
```

Comment:
SRS is a UDP client here, which connect to the peer viewer as a UDP server. If SRS is running in cloud server, and peer viewer is behind NAT, so SRS is not able to connect to the peer viewer.

```
peer viewer ---NAT--> Router ---NAT---> Router --> Cloud VPC ---> SRS
```